modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java

```diff
@@ -616,12 +625,80 @@ else if (log.isInfoEnabled()) {
         if (!res && node.clientRouterNodeId() == null && nodeAlive(nodeId)) {
             LT.warn(log, "Failed to ping node (status check will be initiated): " + nodeId);
 
-            msgWorker.addMessage(new TcpDiscoveryStatusCheckMessage(locNode, node.id()));
+            msgWorker.addMessage(createTcpDiscoveryStatusCheckMessage(locNode, locNode.id(), node.id()));
         }
 
         return res;
     }
 
+    /**
+     * Creates new instance of {@link TcpDiscoveryStatusCheckMessage} trying to choose most optimal constructor.
+     *
+     * @param creatorNode Creator node.
+     * @param creatorNodeId Creator node id, is used when <code>creatorNode</code> is null.
+     * @param failedNodeId Failed node id.
+     * @return new instance of {@link TcpDiscoveryStatusCheckMessage}.
+     */
+    private TcpDiscoveryStatusCheckMessage createTcpDiscoveryStatusCheckMessage(
+        @Nullable TcpDiscoveryNode creatorNode,
+        @Nullable UUID creatorNodeId,
+        UUID failedNodeId
+    ) {
+        assert creatorNode != null || creatorNodeId != null;
+
+        TcpDiscoveryStatusCheckMessage msg;
+
+        if (clusterSupportsTcpDiscoveryNodeSerializationOptimization) {
+            TcpDiscoveryNode crd = resolveCoordinator();
+
+            boolean sameMacs = creatorNode != null && crd != null && U.sameMacs(creatorNode, crd);
+
+            msg = new TcpDiscoveryStatusCheckMessage(
+                creatorNode == null ? creatorNodeId : creatorNode.id(),
+                spi.getNodeAddresses(creatorNode, sameMacs),
+                failedNodeId
+            );
+        }
+        else {
+            if (creatorNode != null)
+                msg = new TcpDiscoveryStatusCheckMessage(creatorNode, failedNodeId);
+            else
+                throw new IgniteSpiException("Cannot create status check message: creator node is unknown.");
+        }
+
+        return msg;
+    }
+
+    /**
+     * Creates new instance of {@link TcpDiscoveryDuplicateIdMessage}.
+     *
+     * @param msg Message.
+     * @return new instance of {@link TcpDiscoveryDuplicateIdMessage}.
+     */
+    private TcpDiscoveryStatusCheckMessage createTcpDiscoveryStatusCheckMessage(
```

Comment:
Not sure we need this method, it doesn't contain any additional logic around call to the method doing actual work.
Please replace with "real" method at the callsite.